userdocs/diagnostics/protocol-type-non-conformance.md

```diff
@@ -0,0 +1,41 @@
+# Protocol type not conforming to itself
+Swift disallows us from using a protocol as a type that conforms to itself as illustrated in the examples below
+
+```swift
+protocol SomeProtocol {
+    init()
+}
+
+struct SomeStruct: SomeProtocol {}
+struct AnotherStruct: SomeProtocol {}
+
+var arr: [SomeProtocol] = [SomeStruct(), AnotherStruct()]
+arr.appendNewItem()
+
+extension Array where Element: SomeProtocol {
+    mutating func appendNewItem() {
+        append(Element())
+    }
+}
+```
+
+The code snippet above would not compile because we are using `SomeProtocol` as a type that conforms to itself. There is no concrete implementation for the protocol.
+
+Consider also the case of using protocol as a type in a generic type - 
+
+```swift
+protocol AnotherProtocol {
+  static func foo()
+}
+
+struct GenericStruct<T: AnotherProtocol> {
+    func faz() {
+        T.foo()
+    }
+}
+
+GenericStruct<AnotherProtocol>().faz()
+```
+Constructing the instance of the struct `GenericStruct` with type `AnotherProtocol` will not compile because there is no concrete implementation for the static requirement of the protocol. 
+There is no implementation for for() used above.
```

Comment:
```suggestion
There is no implementation for `foo()` used above.
```

I think there's a rich set of real-world examples in the Swift Forums that users have stumbled on. Those can often be more valuable for teaching than a contrived example. I've been told that Swift documentation tries to avoid placeholder names in code examples like `i`, `x`, `foo`, instead preferring something more descriptive. Here, though, there's no semantics to be more descriptive about; I think it'd be much easier if we started from a real-world example someone has actually had trouble with, since any of those APIs would have real names and real semantics.

One simple example is probably enough, if it's well chosen. Having illustrated and described the issue, though (as I said to @owenv about the note on PATs), it can be helpful to show (at least briefly) what users who encounter this error should do _instead_. After all, they've got an error in their code, and they're here because they don't fully understand it. Here are some possible suggestions, but we should look to what the real-world examples tend to look like and try to see which ones are most frequently applicable:

In the case of an array, some users could use a homogeneous array instead; perhaps it'd have to be of type `[Any]`, though. In some cases, users could (or even should) use generics. In the general case, however, users would need to create a manual type-eraser type; _how_ to implement such a type would firmly be outside the scope of this note, however.